### PR TITLE
feat(dashboards): Add fields for storing top-level filters

### PR DESCRIPTION
**Pull request by narsaynorath**
We need to be able to pin filters such as projects, environments, and time range (as well as a future `releases`, and `transactions` filter). This PR adds both `projects` (as an M2M relationship since it eases handling deletion of projects from saved dashboard data) and `filters` which is a JSONField to capture the remaining filters like environment and time range.

This approach is similar to how `DiscoverSavedQueries` handles saved filters.

### CODE REVIEW

File: src/sentry/models/dashboard.py [hunk 1]
```diff
@@ -3,6 +3,19 @@
 
 from sentry.db.models import FlexibleForeignKey, Model, sane_repr
 from sentry.db.models.fields.bounded import BoundedBigIntegerField
+from sentry.db.models.fields.jsonfield import JSONField
+
+
+class DashboardProject(Model):
```
Comment (narsaynorath) [inline src/sentry/models/dashboard.py:9]: Is it necessary to explicitly define a through table? I know Django can auto-create one for you, but I noticed our M2M fields always define a table explicitly.

Is it to be more explicit in controlling values such as table name and constraints?

Comment (markstory) [reply src/sentry/models/dashboard.py:9]: > Is it to be more explicit in controlling values such as table name and constraints? 

Partly this, and partly a convention from old django.

Comment (markstory) [reply src/sentry/models/dashboard.py:9]: If we're not placing any logic on the junction table and we can get the table name conventions on the generated model, I don't see a reason to have a formally defined junction model.

Comment (wedamija) [reply src/sentry/models/dashboard.py]: One useful thing with declaring the table explicitly is that you can add a `date_added` column to it, which is sometimes useful for investigations.

**Review by wedamija — APPROVED** [review]
lgtm, any reason we don't make a M2M to environments instead of using filters?

**Review by shruthilayaj — APPROVED** [review]

**Comment by narsaynorath** [comment]
> lgtm, any reason we don't make a M2M to environments instead of using filters?

@wedamija I was told that we don't support deleting environments from the app, so it wasn't as important to use a M2M to store that relationship

**Pull request merged**